File: app/src/main/kotlin/tech/mmarca/openvitals/features/activity/maps/OfflineMapMetadataStore.kt
package tech.mmarca.openvitals.features.activity.maps

import java.io.File
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import kotlinx.serialization.json.put

internal class OfflineMapMetadataStore(
    private val metadataFile: File,
    private val mapsDirectory: File,
) {
    fun read(): OfflineMapLibraryState {
        val root = runCatching {
            Json.parseToJsonElement(metadataFile.readText()).jsonObject
        }.getOrNull() ?: return OfflineMapLibraryState()

        val packs = root[PacksKey]
            ?.let { element -> runCatching { element.jsonArray }.getOrNull() }
            ?.mapNotNull { element -> runCatching { element.jsonObject.toMapPack() }.getOrNull() }
            ?.filter { pack -> pack.file.exists() }
            .orEmpty()
            .sortedByDescending { it.importedAtMillis }

        val activeFormat = root.string(ActiveFormatKey)
            ?.let { value -> runCatching { OfflineMapPackFormat.valueOf(value) }.getOrNull() }
            ?.takeIf { format -> packs.any { it.format == format } }
            ?: root.string(ActiveMapIdKey)
                ?.let { activeId -> packs.firstOrNull { it.id == activeId }?.format }

        return OfflineMapLibraryState(
            mapPacks = packs,
            activeFormat = activeFormat,
        )
    }

    fun write(state: OfflineMapLibraryState) {
        mapsDirectory.mkdirs()
        val normalizedActiveFormat = state.activeFormat
            ?.takeIf { format -> state.mapPacks.any { it.format == format } }
        val json = buildJsonObject {
            if (normalizedActiveFormat != null) {
                put(ActiveFormatKey, normalizedActiveFormat.name)
            }
            put(PacksKey, state.mapPacks.toJsonArray())
        }.toString()

        val tempFile = File(metadataFile.parentFile ?: mapsDirectory, "${metadataFile.name}.tmp")
        tempFile.writeText(json)
        if (!tempFile.renameTo(metadataFile)) {
            tempFile.copyTo(metadataFile, overwrite = true)
            tempFile.delete()
        }
    }

    private fun JsonObject.toMapPack(): OfflineMapPack? {
        val id = string(IdKey)?.takeIf { it.isNotBlank() } ?: return null
        val displayName = string(DisplayNameKey)?.takeIf { it.isNotBlank() } ?: return null
        val savedOriginalFileName = string(OriginalFileNameKey)?.takeIf { it.isNotBlank() }
        val format = string(FormatKey)
            ?.let { value -> runCatching { OfflineMapPackFormat.valueOf(value) }.getOrNull() }
            ?: savedOriginalFileName?.let(OfflineMapPackFormat::fromFileName)
            ?: OfflineMapPackFormat.PMTILES
        val originalFileName = savedOriginalFileName ?: "$displayName${format.fileExtension}"
        val sizeBytes = long(SizeBytesKey)?.takeIf { it >= 0L } ?: 0L
        val importedAtMillis = long(ImportedAtMillisKey)?.takeIf { it > 0L } ?: 0L
        return OfflineMapPack(
            id = id,
            displayName = displayName,
            originalFileName = originalFileName,
            sizeBytes = sizeBytes,
            importedAtMillis = importedAtMillis,
            path = File(mapsDirectory, "$id${format.fileExtension}").absolutePath,
            format = format,
        )
    }

    private fun List<OfflineMapPack>.toJsonArray(): JsonArray =
        buildJsonArray {
            forEach { pack ->
                add(
                    buildJsonObject {
                        put(IdKey, pack.id)
                        put(DisplayNameKey, pack.displayName)
                        put(OriginalFileNameKey, pack.originalFileName)
                        put(FormatKey, pack.format.name)
                        put(SizeBytesKey, pack.sizeBytes)
                        put(ImportedAtMillisKey, pack.importedAtMillis)
                    },
                )
            }
        }

    private fun JsonObject.string(key: String): String? =
        this[key]?.jsonPrimitive?.contentOrNull

    private fun JsonObject.long(key: String): Long? =
        this[key]?.jsonPrimitive?.longOrNull

    private companion object {
        const val ActiveFormatKey = "activeFormat"
        const val ActiveMapIdKey = "activeMapId"
        const val PacksKey = "packs"
        const val IdKey = "id"
        const val DisplayNameKey = "displayName"
        const val OriginalFileNameKey = "originalFileName"
        const val FormatKey = "format"
        const val SizeBytesKey = "sizeBytes"
        const val ImportedAtMillisKey = "importedAtMillis"
    }
}

File: lib/features/activity/maps/offline_map_metadata_store.dart
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:shared_preferences/shared_preferences.dart';

import 'offline_map_models.dart';

/// Persists the offline map library, ported from the Kotlin
/// `OfflineMapMetadataStore`.
///
/// The Kotlin store serialises to a `metadata.json` file; the Dart port keeps
/// the same JSON shape (so old payloads round-trip, including the legacy
/// `activeMapId` migration) but stores the string wherever [readRaw] /
/// [writeRaw] point — [OfflineMapMetadataStore.sharedPreferences] backs them
/// with [SharedPreferences]. Pack file paths are reconstructed from
/// [mapsDirectoryPath] and packs whose files no longer exist are dropped on
/// read, matching the source. All of this is pure logic: the persistence and
/// file-existence seams are injectable, so a test can round-trip in memory.
class OfflineMapMetadataStore {
  OfflineMapMetadataStore({
    required this.readRaw,
    required this.writeRaw,
    required this.mapsDirectoryPath,
    bool Function(String path)? fileExists,
  }) : _fileExists = fileExists ?? _defaultFileExists;

  /// SharedPreferences-backed store: the JSON payload lives under [prefsKey].
  factory OfflineMapMetadataStore.sharedPreferences(
    SharedPreferences prefs,
    String mapsDirectoryPath, {
    String prefsKey = defaultPrefsKey,
    bool Function(String path)? fileExists,
  }) =>
      OfflineMapMetadataStore(
        readRaw: () => prefs.getString(prefsKey),
        writeRaw: (value) => prefs.setString(prefsKey, value),
        mapsDirectoryPath: mapsDirectoryPath,
        fileExists: fileExists,
      );

  /// Reads the raw persisted JSON payload (null when absent).
  final String? Function() readRaw;

  /// Writes the raw JSON payload.
  final void Function(String value) writeRaw;

  /// Directory the pack files live in; used to reconstruct pack paths.
  final String mapsDirectoryPath;

  final bool Function(String path) _fileExists;

  static const String defaultPrefsKey = 'offline_maps_metadata';

  static bool _defaultFileExists(String path) => File(path).existsSync();

  static const String _activeFormatKey = 'activeFormat';
  static const String _activeMapIdKey = 'activeMapId';
  static const String _packsKey = 'packs';
  static const String _idKey = 'id';
  static const String _displayNameKey = 'displayName';
  static const String _originalFileNameKey = 'originalFileName';
  static const String _formatKey = 'format';
  static const String _sizeBytesKey = 'sizeBytes';
  static const String _importedAtMillisKey = 'importedAtMillis';

  OfflineMapLibraryState read() {
    final root = _decodeRoot();
    if (root == null) return const OfflineMapLibraryState();

    final rawPacks = root[_packsKey];
    final packs = <OfflineMapPack>[];
    if (rawPacks is List) {
      for (final element in rawPacks) {
        if (element is! Map) continue;
        final pack = _toMapPack(element.cast<String, dynamic>());
        if (pack != null && _fileExists(pack.path)) {
          packs.add(pack);
        }
      }
    }
    packs.sort((a, b) => b.importedAtMillis.compareTo(a.importedAtMillis));

    final activeFormat = _resolveActiveFormat(root, packs);
    return OfflineMapLibraryState(mapPacks: packs, activeFormat: activeFormat);
  }

  void write(OfflineMapLibraryState state) {
    final normalizedActiveFormat = state.activeFormat != null &&
            state.mapPacks.any((pack) => pack.format == state.activeFormat)
        ? state.activeFormat
        : null;
    final json = <String, dynamic>{
      if (normalizedActiveFormat != null)
        _activeFormatKey: normalizedActiveFormat.storageName,
      _packsKey: state.mapPacks
          .map(
            (pack) => <String, dynamic>{
              _idKey: pack.id,
              _displayNameKey: pack.displayName,
              _originalFileNameKey: pack.originalFileName,
              _formatKey: pack.format.storageName,
              _sizeBytesKey: pack.sizeBytes,
              _importedAtMillisKey: pack.importedAtMillis,
            },
          )
          .toList(),
    };
    writeRaw(jsonEncode(json));
  }

  Map<String, dynamic>? _decodeRoot() {
    final raw = readRaw();
    if (raw == null || raw.isEmpty) return null;
    try {
      final decoded = jsonDecode(raw);
      return decoded is Map ? decoded.cast<String, dynamic>() : null;
    } catch (_) {
      return null;
    }
  }

  OfflineMapPack? _toMapPack(Map<String, dynamic> json) {
    final id = _stringOrNull(json[_idKey]);
    if (id == null || id.isEmpty) return null;
    final displayName = _stringOrNull(json[_displayNameKey]);
    if (displayName == null || displayName.isEmpty) return null;

    final savedOriginalFileName = _stringOrNull(json[_originalFileNameKey]);
    final format = _formatOrNull(json[_formatKey]) ??
        (savedOriginalFileName != null
            ? OfflineMapPackFormat.fromFileName(savedOriginalFileName)
            : null) ??
        OfflineMapPackFormat.pmtiles;
    final originalFileName =
        (savedOriginalFileName != null && savedOriginalFileName.isNotEmpty)
            ? savedOriginalFileName
            : '$displayName${format.fileExtension}';
    final sizeBytes = _intOrNull(json[_sizeBytesKey]);
    final importedAtMillis = _intOrNull(json[_importedAtMillisKey]);

    return OfflineMapPack(
      id: id,
      displayName: displayName,
      originalFileName: originalFileName,
      sizeBytes: (sizeBytes != null && sizeBytes >= 0) ? sizeBytes : 0,
      importedAtMillis:
          (importedAtMillis != null && importedAtMillis > 0) ? importedAtMillis : 0,
      path: p.join(mapsDirectoryPath, '$id${format.fileExtension}'),
      format: format,
    );
  }

  OfflineMapPackFormat? _resolveActiveFormat(
    Map<String, dynamic> root,
    List<OfflineMapPack> packs,
  ) {
    final rawFormat = _formatOrNull(root[_activeFormatKey]);
    if (rawFormat != null && packs.any((pack) => pack.format == rawFormat)) {
      return rawFormat;
    }
    final legacyId = _stringOrNull(root[_activeMapIdKey]);
    if (legacyId != null) {
      for (final pack in packs) {
        if (pack.id == legacyId) return pack.format;
      }
    }
    return null;
  }

  static OfflineMapPackFormat? _formatOrNull(Object? value) {
    final name = _stringOrNull(value);
    return name == null ? null : OfflineMapPackFormat.fromStorage(name);
  }

  static String? _stringOrNull(Object? value) => value is String ? value : null;

  static int? _intOrNull(Object? value) {
    if (value is int) return value;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value);
    return null;
  }
}
